### ISSUE TITLE
Async.AwaitTask increases stack, causing StackOverflowException

### ISSUE THREAD

**Issue by Thorium** [dotnet/fsharp#12749]
Doing code that looks like tail-recursive, using `Async.AwaitTask` may cause surprising StackOverflowException

**Repro steps**

Here is a code example, that looks tail-recursive:
```fsharp
// Example 1
let work =
    // simulate some C# library call returning a task
    System.Threading.Tasks.Task.Run(fun () -> 1) 
let rec asdfasdf acc =
    async {
        match acc with
        | 5000 -> return "done"
        | _ ->
            let! c = work |> Async.AwaitTask
            return! asdfasdf (acc+c)
    }
asdfasdf 0 |> Async.RunSynchronously
// Process is terminated due to StackOverflowException.
```

The stack is well hidden by Visual Studio so that it's hard to spot the reason.
Here is a non-functional way to encounter the same issue:

```fsharp
// Example 2
let work =
    // simulate some C# library call returning a task
    System.Threading.Tasks.Task.Run(fun () -> 1) 
async {
    let mutable acc = 1
    while acc < 5000 do
        let! c = work |> Async.AwaitTask
        acc <- acc + c
    return "done"
} |> Async.RunSynchronously
// Process is terminated due to StackOverflowException.
```

**Expected behavior**

It would be quite convinient to not cause a StackOverflowException.

**Actual behavior**

StackOverflowException.

**Known workarounds**

- I don't know any easy way to avoid this in F# 5, at least when the tasks are somewhat related so that they cannot be executed in parallel.
- Don't do async is a weak work-around when you are in production web-server with hundreds of users. The webserver will just run out of threads.
- Maybe there would be some external libraries with async-seq, that could help. Adding more dependencies.

**Related information**

I expect this happens on any environment, but here are my machine details:

* Windows 10
* .NET Framework 4.7.2
* Microsoft (R) F# Interactive version 11.4.2.0 for F# 5.0, FSharp.Core, Version=5.0.0.0

**Comment by Thorium** [comment]
I found a workaround, avoiding `AwaitTask` and using `AwaitIAsyncResult` instead:

```fsharp
// Example 1
let work =
    // simulate some C# library call returning a task
    System.Threading.Tasks.Task.Run(fun () -> 1) 
let rec asdfasdf acc =
    async {
        match acc with
        | 5000 -> return "done"
        | _ ->
            let! c = work |> Async.AwaitIAsyncResult
            if work.IsFaulted then raise work.Exception
            return! asdfasdf (acc+work.Result)
    }
asdfasdf 0 |> Async.RunSynchronously
```
and
```fsharp
let work =
    // simulate some C# library call returning a task
    System.Threading.Tasks.Task.Run(fun () -> 1) 
async {
    let mutable acc = 1
    while acc < 5000 do
        let! c = work |> Async.AwaitIAsyncResult
        if work.IsFaulted then raise work.Exception
        acc <- acc + work.Result
    return "done"
} |> Async.RunSynchronously
```

**Comment by dsyme** [comment]
The original is by-design - AwaitTask on a synchronous task  continues on the same thread and can consumes stack as it calls its continuation.  

I believe there are several workarounds such as adding an `Async.Sleep(0)` or some similar change to the otherwise synchronous loop.   Using AwaitIAsyncResult has a similar effect.

**Comment by Thorium** [comment]
Confirmed that the issue is fixed in F# 7.0, FSharp.Core 7.0.200
